monitor.go

```diff
@@ -0,0 +1,223 @@
+package main
+
+import (
+	"encoding/hex"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+	raven "github.com/getsentry/raven-go"
+	"github.com/pborman/uuid"
+	"github.com/taskcluster/statsum"
+	tcauth "github.com/taskcluster/taskcluster-client-go/auth"
+)
+
+// NewMonitor creates a new monitor
+func NewMonitor(project string, auth *tcauth.Auth, logLevel string, tags map[string]string) *Monitor {
+	// Create statsumConfigurer
+	statsumConfigurer := func(project string) (statsum.Config, error) {
+		res, err := auth.StatsumToken(project)
+		if err != nil {
+			return statsum.Config{}, err
+		}
+		return statsum.Config{
+			Project: res.Project,
+			BaseURL: res.BaseURL,
+			Token:   res.Token,
+			Expires: time.Time(res.Expires),
+		}, nil
+	}
+
+	// Create logger and parse logLevel
+	logger := logrus.New()
+	switch strings.ToLower(logLevel) {
+	case logrus.DebugLevel.String():
+		logger.Level = logrus.DebugLevel
+	case logrus.InfoLevel.String():
+		logger.Level = logrus.InfoLevel
+	case logrus.WarnLevel.String():
+		logger.Level = logrus.WarnLevel
+	case logrus.ErrorLevel.String():
+		logger.Level = logrus.ErrorLevel
+	case logrus.FatalLevel.String():
+		logger.Level = logrus.FatalLevel
+	case logrus.PanicLevel.String():
+		logger.Level = logrus.PanicLevel
+	default:
+		panic(fmt.Sprintf("Unsupported log-level: %s", logLevel))
+	}
+
+	// Convert tags to logrus.Fields
+	fields := make(logrus.Fields, len(tags))
+	for k, v := range tags {
+		fields[k] = v
+	}
+
+	// Declare monitor so we can reference it in OnError
+	var m *Monitor
+	m = &Monitor{
+		Statsum: statsum.New(project, statsumConfigurer, statsum.Options{
+			OnError: func(err error) { m.ReportWarning(err) },
+		}),
+		Entry: logrus.NewEntry(logger).WithFields(fields),
+		sentry: &sentry{
+			client:  nil,
+			project: project,
+			auth:    auth,
+		},
+	}
+
+	return m
+}
+
+type sentry struct {
+	client     *raven.Client
+	m          sync.Mutex
+	project    string
+	expiration time.Time
+	auth       *tcauth.Auth
+}
+
+func (s *sentry) Client() (*raven.Client, error) {
```

Comment:
Here `sentry` is not exported, so `Client()` is not available outside the package, so maybe this should be `client()`? Note, the generic-worker is horribly inconsistent everywhere, but I just wanted to call it out as it seemed strange having an unexported type with an exported method (or is it satisfying an interface?)